### PR TITLE
Add geopackage cache

### PR DESCRIPTION
**Pull request by jsvrcek**
Added a geopackage cache backend, referring to some work S. Lander did.  It uses the MBTiles as a basis, and then just adds the schemas and some checks to verify the datasets.

I also updated the documentation and added a configuration example.

Let me know what you think, or if I am missing something. Thanks.

### CODE REVIEW

**Comment by jj0hns0n** [comment]
@jsvrcek you should probably squash these commits

File: doc/caches.rst [hunk 1]
```diff
@@ -288,3 +288,39 @@ Example
...
             http: 8098
+
+
+``geopackage``
+========
+
+.. versionadded:: 1.8.x
+
+Store tiles in a `geopackage <http://www.geopackage.org/>`_ database. MapProxy creates a tile table if one isn't defined and populates the required meta data fields.
+This backend is good for datasets that require portabilitiy. Data can be distributed over multiple nodes providing a fault-tolernt and high-available storage. A Riak cluster is masterless and each node can handle read and write requests.
```
Comment (olt) [inline doc/caches.rst]: Riak? :-)

File: doc/mapproxy_util_autoconfig.rst [hunk 2]
```diff
@@ -26,8 +26,12 @@ Options
 
 .. cmdoption:: --capabilities <url|filename>
 
-  URL or filename of the WMS capabilities document. The tool will add `REQUEST` and `SERVICE` parameters to the URL as necessary.
+  URL or filename of the WMS capabilities document. The tool will add `REQUEST` and `SERVICE` parameters to the URL as necessary.  This option cannot be used with the --geopackage option.
```
Comment (olt) [inline doc/mapproxy_util_autoconfig.rst]: Can you create a separate pull request for the changes to autoconfig? It's unrelated to the GeoPackage cache backend and makes the pull request harder to review.

File: mapproxy/cache/geopackage.py [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+                 ]
+
+        for stmt in stmts:
+            db.execute(stmt)
+
+        db.execute("PRAGMA foreign_keys = 1;")
+
+        # List of WKT execute statements and data.
+        wkts = [("""
+                INSERT OR REPLACE INTO gpkg_spatial_ref_sys (
```
Comment (olt) [inline mapproxy/cache/geopackage.py]: Can you refactor this SQL statement? The same INSERT INTO lines are repeated 5 times.

File: mapproxy/cache/geopackage.py [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+    @param: String of data to be tested.
+    @return: if data is only alphanumeric or '_' chars.
+    """
+    import re
+    if re.match(r'\w+$', data):
+        return True
+    return False
+
+
+def isclose(a, b, rel_tol=1e-09, abs_tol=0.0):
```
Comment (olt) [inline mapproxy/cache/geopackage.py]: PEP8: `is_close`

File: mapproxy/cache/geopackage.py [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+            log.info("Cursor rowcount = {0}".format(cursor.rowcount))
+            if cursor.rowcount:
+                return True
+            return False
+
+    def load_tile_metadata(self, tile):
+        self.load_tile(tile)
+
+
+class GeopackageLevelCache(TileCacheBase):
```
Comment (olt) [inline mapproxy/cache/geopackage.py]: GeopackageLevelCache is only used in the tests?

File: mapproxy/cache/geopackage.py [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+                ensure_directory(self.geopackage_file)
+                self._initialize_gpkg()
+        else:
+            if not self.check_gpkg():
+                ensure_directory(self.geopackage_file)
+                self._initialize_gpkg()
+
+    def check_gpkg(self):
+        if not self._verify_table():
+            print("table invalid")
```
Comment (olt) [inline mapproxy/cache/geopackage.py]: There are still a few debug prints in this code.

File: mapproxy/cache/geopackage.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+                return False
+            return True
+
+    def _verify_gpkg_contents(self):
+        with sqlite3.connect(self.geopackage_file) as db:
+            try:
+                cur = db.execute("""SELECT table_name, data_type, srs_id FROM gpkg_contents WHERE table_name = ?"""
+                                 , (self.table_name,))
+            except sqlite3.OperationalError as oe:
+                if 'no such column' in str(oe):
```
Comment (olt) [inline mapproxy/cache/geopackage.py]: Is there a better way to do these checks? The message can change with other SQLite versions.

**Comment by jsvrcek** [comment]
Thanks for looking at that, I pulled out the autoconfig stuff for now, I'll submit a new PR for that if you pull this in.  I fixed those things that you mentioned yesterday and the branch should reflect that,  Let me know if there is anything else I missed.  Thanks.

File: mapproxy/test/system/test_cache_geopackage.py [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+
+    shutil.copy(os.path.join(test_config['fixture_dir'], 'cache.gpkg'),
+        test_config['base_dir'])
+    create_app(test_config)
+
+def teardown_module():
+    module_teardown(test_config)
+
+
+class TestGeopackageCache(SystemTest, TileCacheTestBase):
```
Comment (olt) [inline mapproxy/test/system/test_cache_geopackage.py]: Can you move the unit tests to unit/test_cache_geopackage.py (i.e. all tests that do not require a service)?

See
- https://github.com/mapproxy/mapproxy/blob/master/mapproxy/test/unit/test_cache_tile.py#L299
- https://github.com/mapproxy/mapproxy/blob/master/mapproxy/test/unit/test_cache_tile.py#L311
- https://github.com/mapproxy/mapproxy/blob/master/mapproxy/test/unit/test_cache_riak.py#L32
for examples.

File: mapproxy/test/system/test_cache_geopackage.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+                                     configuration.grids.get('GLOBAL_GEODETIC'),
+                                     self.table_name)
+
+    def teardown(self):
+        if self.cache:
+            self.cache.cleanup()
+        TileCacheTestBase.teardown(self)
+
+    def test_get_map_cached(self):
+        prepare_env(test_config, 'cache_geopackage.yaml')
```
Comment (olt) [inline mapproxy/test/system/test_cache_geopackage.py]: `prepare_enc` and `create_app` are "expensive" and should only be called on module or class level. They should also always be followed by a `teardown` call.

Normally, for the system tests there is one Python test file for each YAML configuration. You can work around that by creating a `SystemTest` subclass for each test configuration and add a `setup_class` and `teardown_class` methods that sets everything up. See http://nose.readthedocs.io/en/latest/writing_tests.html#test-classes

But see also my next comment. Maybe multiple classes are not necessary.

File: mapproxy/test/system/test_cache_geopackage.py [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+        eq_(resp.content_type, 'image/png')
+        data = BytesIO(resp.body)
+        assert is_png(data)
+
+    def test_bad_config_geopackage_srs(self):
+        error_msg = None
+        try:
+            prepare_env(test_config, 'cache_geopackage_srs.yaml')
+            create_app(test_config)
+        except ValueError as ve:
```
Comment (olt) [inline mapproxy/test/system/test_cache_geopackage.py]: I think system tests are a bit overkill for this. I would create a few unit tests where you create GeopackageCache objects with varying tile grids.

File: mapproxy/cache/geopackage.py [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+from mapproxy.srs import get_epsg_num
+
+import logging
+
+log = logging.getLogger(__name__)
+
+class GeopackageCache(TileCacheBase):
+    supports_timestamp = False
+
+    def __init__(self, geopackage_file, grid_conf, table_name, with_timestamps=False):
```
Comment (olt) [inline mapproxy/cache/geopackage.py]: GeopackageCache should take a TileGrid and not a grid_conf.

**Comment by jsvrcek** [comment]
I made those changes, I hope that is what you intended.  The one thing I wasn't sure you would support (because no unit tests have fixtures already), was I copied the system tests fixture into the unit tests.  I didn't want the unit tests to use fixtures from system, but I know it is redundant.  Let me know if you would like me to do it a different way.   Thanks.

**Comment by olt** [comment]
Thanks a lot for your work.
`TileCacheTestBase` is meant for unit tests and I moved these tests back to `unit/`.
I also ported my recet `bulk_meta_tile` improvements for MBTiles back to the GeoPackage cache.

The changes are rebased and merged into master.

**Comment by olt** [comment]
If anyone is using this from the master branch: I just changed the `dirname` option to `directory` and the default `table_name` to `cachename_gridname` instead of `cachename_epsgcode`, to be more consistent with the other caches.

cc @jsvrcek

**Pull request closed without merging**